### PR TITLE
feat: Allow filtering projects by project status

### PR DESCRIPTION
**Pull request by joshua-mo-143**
* Adds a flag to allow filtering projects by deployment status.

Ref: Issue #612 

Notes:
Please ignore logout function removal - this is due to incorrect branching from an earlier PR unfortunately, some caution may not have been exercised 😅

### CODE REVIEW

File: common/src/models/project.rs [hunk 1]
```diff
@@ -57,13 +57,22 @@ pub struct AdminResponse {
...
 
-pub fn get_table(projects: &Vec<Response>) -> String {
+pub fn get_table(projects: &Vec<Response>, filter: Option<State>) -> String {
     if projects.is_empty() {
         format!(
             "{}\n",
             "No projects are linked to this account".yellow().bold()
         )
     } else {
+        let projects = match filter {
```
Comment (oddgrd) [inline common/src/models/project.rs:67]: So currently all projects are fetched, then filtered. We should rather apply the filter when fetching the projects from the database, and then only send the filtered projects to the client. We could add a route to the gateway API for this, with the desired state passed in as a path parameter, so something like "/projects/:state". And then make a handler that selects from the db the projects owned by this user with the desired state.

**Review by oddgrd — COMMENTED** [review]
Cool! Thanks josh. I added some comments about how I think we could make this a bit more efficient. I'd also like to point out that this is filtering by project status, not deployment status, but I realize it is a bit confusing :smile: The project status indicates the status of the deployer container, while the deployment status indicates the status of the deployed service within.

I'm also wondering if we should be able to filter by more than those three states, but I'll have to get back to you on that.

File: cargo-shuttle/src/args.rs [hunk 2]
```diff
@@ -108,7 +108,11 @@ pub enum ProjectCommand {
     /// create an environment for this project on shuttle
     New,
     /// list all projects belonging to the calling account
-    List,
+    List {
+        #[arg(long)]
+        /// Follow status of project command
```
[the remark this answers is not in the document]
Comment (joshua-mo-143) [reply cargo-shuttle/src/args.rs]: Amended in new commit

File: common/src/models/project.rs (hunk 1 above)
Comment (joshua-mo-143) [reply common/src/models/project.rs]: I've pushed a new commit to try and amend this.

**Review by oddgrd — APPROVED** [review]
Nice! This looks good to go, but would you mind adding a couple of assertions next to the `iter_user_projects_detailed` test? See https://github.com/shuttle-hq/shuttle/blob/main/gateway/src/service.rs#L709-L716

**Comment by joshua-mo-143** [comment]
No worries - I'm at work at the moment but this will be done by end of tomorrow or over the weekend if not tomorrow.

File: gateway/src/service.rs [hunk 3]
```diff
@@ -295,6 +295,27 @@ impl GatewayService {
         Ok(iter)
     }
 
+    pub async fn iter_user_projects_detailed_filtered(
+        &self,
+        account_name: AccountName,
+        filter: String,
+    ) -> Result<impl Iterator<Item = (ProjectName, Project)>, Error> {
+        let iter =
+            query("SELECT project_name, project_state FROM projects WHERE account_name = ?1 AND WHERE project_state = ?2")
```
[the remark this answers is not in the document]
Comment (joshua-mo-143) [reply gateway/src/service.rs]: Amended

**Comment by joshua-mo-143** [comment]
Accidentally pushed too early - please ignore 🤦

**Comment by joshua-mo-143** [comment]
Should be ready to review now, some relevant assertions were added. Let me know if anything needs changing before merging

**Comment by oddgrd** [comment]
LGTM, thanks! :partying_face:

**Pull request merged**